### DISCUSSION TITLE
flush head element to trigger static assets loading with streaming before the loader resolved

### DISCUSSION THREAD

**Discussion by lili21** [remix-run/remix#4805] [Proposals]
stylesheet will blocking the page rendering. we should make it loading as quick as possible. 

we could get all the links which export from route module, I guess we can and should flush it before the loader resolve.

**Comment by lili21** [comment]
same rule for the js assets.

**Comment by richardscarrott** [comment]
I just came here to propose this too 👍

As you suggest, it looks although `handleRequest` in `server.entry.ts` isn't called until after `loaders` have resolved so there's a period of time where the browser is just sitting there waiting, even though we know it'll eventually need to go and fetch a bunch of static resources (CSS, Images, JS etc.) -- some of which are render blocking.

Given Remix can synchronously determine which static resources are required for any given route, it could offer significant performance gain if they're flushed to the client before loaders have completed.

I think this could be done v. nicely via 103 early hints, however last time I checked they have almost no support across popular servers.

If that's not an option then, given the HTML is already being streamed for non-bots in the default template (i.e. an accurate http status code is forgone), the head could be flushed early with a 200.

It could be tricky in Remix as the head is rendered as part of a single React render. I expect the root could be rendered separately although prob not ideal for managing head tags once in the client, or maybe React Suspense + SSR streaming could help here, e.g.

```ts
// Render this before `Loaders` have finished?
const Root = () => (
  <html>
    <head>
        <Links />
        <Scripts preload />
    </head>
    <Suspense fallback={null}>
       <AwaitLoaders>
         <body>
           <Outlet />
         </body>
       </AwaitLoaders>
    <Suspense>
  </html>
);
```

The downside of using React SSR Streaming is I think it requires JS which is a shame.

EDIT: I just realised loaders can redirect / set cookies etc. so flushing the head early prob isn't viable because that behaviour would have to be replicated in JS which is a PITA. (i.e. `res.write('<script>location.href = ${redirectUrl}</script>')`)

**Reply by lili21** [reply]
You are read my mind 👍 . 
server push is another option, but chrome is [removing it](https://developer.chrome.com/blog/removing-push/)

**Reply by richardscarrott** [reply]
Oh interesting. I just made an edit as I realised loaders can redirect / set cookies which prob makes flushing the head with a 200 unrealistic. I think it might therefore just be a case of waiting for wider 103 early hints support 😞

**Comment by ryanflorence** [comment]
You will be able to do this yourself when we ship `defer`. I'd be uncomfortable ever doing it by default until we have [103 EARLY HINTS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103) because you lose the ability to send a proper status code.
